File: apps/demo-be-fsharp-giraffe/src/DemoBeFsgi/Handlers/AuthHandler.fs
module DemoBeFsgi.Handlers.AuthHandler

open System
open System.Text.Json
open Giraffe
open DemoBeFsgi.Infrastructure.AppDbContext
open DemoBeFsgi.Infrastructure.PasswordHasher
open DemoBeFsgi.Infrastructure.Repositories.RepositoryTypes
open DemoBeFsgi.Domain.User
open DemoBeFsgi.Domain.Types
open DemoBeFsgi.Auth.JwtService
open DemoBeFsgi.Contracts.ContractWrappers

let private maxFailedAttempts = 5

let register: HttpHandler =
    fun _next ctx ->
        task {
            let! body = ctx.ReadBodyFromRequestAsync()

            let req =
                try
                    JsonSerializer.Deserialize<RegisterRequest>(
                        body,
                        JsonSerializerOptions(PropertyNameCaseInsensitive = true)
                    )
                    |> Some
                with _ ->
                    None

            match req with
            | None ->
                ctx.Response.StatusCode <- 400

                return!
                    json
                        {| error = "Bad Request"
                           message = "Invalid request body" |}
                        earlyReturn
                        ctx
            | Some r ->
                let usernameResult = validateUsername (if r.username = null then "" else r.username)
                let emailResult = validateEmail (if r.email = null then "" else r.email)
                let passwordResult = validatePassword (if r.password = null then "" else r.password)

                match usernameResult, emailResult, passwordResult with
                | Error(ValidationError(f, m)), _, _ ->
                    ctx.Response.StatusCode <- 400

                    return!
                        json
                            {| error = "Validation Error"
                               field = f
                               message = m |}
                            earlyReturn
                            ctx
                | _, Error(ValidationError(f, m)), _ ->
                    ctx.Response.StatusCode <- 400

                    return!
                        json
                            {| error = "Validation Error"
                               field = f
                               message = m |}
                            earlyReturn
                            ctx
                | _, _, Error(ValidationError(f, m)) ->
                    ctx.Response.StatusCode <- 400

                    return!
                        json
                            {| error = "Validation Error"
                               field = f
                               message = m |}
                            earlyReturn
                            ctx
                | Ok _, Ok _, Ok _ ->
                    let userRepo = ctx.GetService<UserRepository>()

                    let! existingUser = userRepo.FindByUsername r.username

                    if existingUser.IsSome then
                        ctx.Response.StatusCode <- 409

                        return!
                            json
                                {| error = "Conflict"
                                   message = "Username already exists" |}
                                earlyReturn
                                ctx
                    else
                        let now = DateTime.UtcNow
                        let userId = Guid.NewGuid()

                        let entity: UserEntity =
                            { Id = userId
                              Username = r.username
                              Email = r.email
                              DisplayName = r.username
                              PasswordHash = hashPassword r.password
                              Role = roleToString User
                              Status = statusToString Active
                              FailedLoginAttempts = 0
                              CreatedAt = now
                              UpdatedAt = now }

                        let! _ = userRepo.Create entity

                        ctx.Response.StatusCode <- 201

                        return!
                            json
                                {| id = userId
                                   username = entity.Username
                                   email = entity.Email
                                   displayName = entity.DisplayName |}
                                earlyReturn
                                ctx
                | _ ->
                    ctx.Response.StatusCode <- 400

                    return!
                        json
                            {| error = "Bad Request"
                               message = "Validation failed" |}
                            earlyReturn
                            ctx
        }

let login: HttpHandler =
    fun _next ctx ->
        task {
            let! body = ctx.ReadBodyFromRequestAsync()

            let req =
                try
                    JsonSerializer.Deserialize<LoginRequest>(
                        body,
                        JsonSerializerOptions(PropertyNameCaseInsensitive = true)
                    )
                    |> Some
                with _ ->
                    None

            match req with
            | None ->
                ctx.Response.StatusCode <- 400

                return!
                    json
                        {| error = "Bad Request"
                           message = "Invalid request body" |}
                        earlyReturn
                        ctx
            | Some r ->
                let userRepo = ctx.GetService<UserRepository>()
                let rtRepo = ctx.GetService<RefreshTokenRepository>()

                let! userOpt = userRepo.FindByUsername r.username

                match userOpt with
                | None ->
                    ctx.Response.StatusCode <- 401

                    return!
                        json
                            {| error = "Unauthorized"
                               message = "Invalid credentials" |}
                            earlyReturn
                            ctx
                | Some user when user.Status = statusToString Locked ->
                    ctx.Response.StatusCode <- 401

                    return!
                        json
                            {| error = "Unauthorized"
                               message = "Account is locked after too many failed attempts" |}
                            earlyReturn
                            ctx
                | Some user when user.Status = statusToString Inactive ->
                    ctx.Response.StatusCode <- 401

                    return!
                        json
                            {| error = "Unauthorized"
                               message = "Account has been deactivated" |}
                            earlyReturn
                            ctx
                | Some user when user.Status = statusToString Disabled ->
                    ctx.Response.StatusCode <- 401

                    return!
                        json
                            {| error = "Unauthorized"
                               message = "Account has been disabled" |}
                            earlyReturn
                            ctx
                | Some user when not (verifyPassword r.password user.PasswordHash) ->
                    let newAttempts = user.FailedLoginAttempts + 1

                    let newStatus =
                        if newAttempts >= maxFailedAttempts then
                            statusToString Locked
                        else
                            user.Status

                    let updated =
                        { user with
                            FailedLoginAttempts = newAttempts
                            Status = newStatus
                            UpdatedAt = DateTime.UtcNow }

                    let! _ = userRepo.Update updated

                    if newAttempts >= maxFailedAttempts then
                        ctx.Response.StatusCode <- 401

                        return!
                            json
                                {| error = "Unauthorized"
                                   message = "Account is locked after too many failed attempts" |}
                                earlyReturn
                                ctx
                    else
                        ctx.Response.StatusCode <- 401

                        return!
                            json
                                {| error = "Unauthorized"
                                   message = "Invalid credentials" |}
                                earlyReturn
                                ctx
                | Some user ->
                    let updated =
                        { user with
                            FailedLoginAttempts = 0
                            UpdatedAt = DateTime.UtcNow }

                    let! _ = userRepo.Update updated

                    let accessToken = generateAccessToken user.Id user.Username user.Email user.Role
                    let refreshTokenStr = generateRefreshToken user.Id

                    let now = DateTime.UtcNow

                    let rtEntity: RefreshTokenEntity =
                        { Id = Guid.NewGuid()
                          UserId = user.Id
                          TokenHash = refreshTokenStr
                          ExpiresAt = now.AddDays(7.0)
                          CreatedAt = now
                          Revoked = false }

                    let! _ = rtRepo.Create rtEntity

                    return!
                        json
                            {| accessToken = accessToken
                               refreshToken = refreshTokenStr
                               tokenType = "Bearer" |}
                            earlyReturn
                            ctx
        }

let refresh: HttpHandler =
    fun _next ctx ->
        task {
            let! body = ctx.ReadBodyFromRequestAsync()

            let req =
                try
                    JsonSerializer.Deserialize<RefreshRequest>(
                        body,
                        JsonSerializerOptions(PropertyNameCaseInsensitive = true)
                    )
                    |> Some
                with _ ->
                    None

            match req with
            | None ->
                ctx.Response.StatusCode <- 400

                return!
                    json
                        {| error = "Bad Request"
                           message = "Invalid request body" |}
                        earlyReturn
                        ctx
            | Some r ->
                let userRepo = ctx.GetService<UserRepository>()
                let rtRepo = ctx.GetService<RefreshTokenRepository>()

                let! rtEntityOpt = rtRepo.FindActiveByHash r.refreshToken

                match rtEntityOpt with
                | None ->
                    ctx.Response.StatusCode <- 401

                    return!
                        json
                            {| error = "Unauthorized"
                               message = "Invalid or already used token" |}
                            earlyReturn
                            ctx
                | Some rtEntity when rtEntity.ExpiresAt < DateTime.UtcNow ->
                    ctx.Response.StatusCode <- 401

                    return!
                        json
                            {| error = "Unauthorized"
                               message = "Token has expired" |}
                            earlyReturn
                            ctx
                | Some rtEntity ->
                    let! userOpt = userRepo.FindById rtEntity.UserId

                    match userOpt with
                    | None ->
                        ctx.Response.StatusCode <- 401

                        return!
                            json
                                {| error = "Unauthorized"
                                   message = "User not found" |}
                                earlyReturn
                                ctx
                    | Some user when user.Status <> statusToString Active ->
                        ctx.Response.StatusCode <- 401

                        return!
                            json
                                {| error = "Unauthorized"
                                   message = "Account has been deactivated" |}
                                earlyReturn
                                ctx
                    | Some user ->
                        let revokedRt = { rtEntity with Revoked = true }
                        let! _ = rtRepo.Update revokedRt

                        let accessToken = generateAccessToken user.Id user.Username user.Email user.Role
                        let newRefreshToken = generateRefreshToken user.Id

                        let now = DateTime.UtcNow

                        let newRtEntity: RefreshTokenEntity =
                            { Id = Guid.NewGuid()
                              UserId = user.Id
                              TokenHash = newRefreshToken
                              ExpiresAt = now.AddDays(7.0)
                              CreatedAt = now
                              Revoked = false }

                        let! _ = rtRepo.Create newRtEntity

                        return!
                            json
                                {| accessToken = accessToken
                                   refreshToken = newRefreshToken
                                   tokenType = "Bearer" |}
                                earlyReturn
                                ctx
        }

let logout: HttpHandler =
    fun next ctx ->
        task {
            let authHeader = ctx.Request.Headers["Authorization"].ToString()

            let token =
                if authHeader.StartsWith("Bearer ", StringComparison.Ordinal) then
                    authHeader.Substring(7)
                else
                    ""

            let tokenRepo = ctx.GetService<TokenRepository>()
            let jti = getTokenJti token

            match jti with
            | Some j ->
                let! exists = tokenRepo.ExistsJti j

                if not exists then
                    let userId =
                        if ctx.Items.ContainsKey("UserId") then
                            ctx.Items["UserId"] :?> Guid
                        else
                            Guid.Empty

                    let revokedEntity: RevokedTokenEntity =
                        { Id = Guid.NewGuid()
                          Jti = j
                          UserId = userId
                          RevokedAt = DateTime.UtcNow }

                    let! _ = tokenRepo.Create revokedEntity
                    ()
            | None -> ()

            return! json {| message = "Logged out successfully" |} next ctx
        }

let logoutAll: HttpHandler =
    fun next ctx ->
        task {
            let tokenRepo = ctx.GetService<TokenRepository>()
            let rtRepo = ctx.GetService<RefreshTokenRepository>()
            let authHeader = ctx.Request.Headers["Authorization"].ToString()

            let token =
                if authHeader.StartsWith("Bearer ", StringComparison.Ordinal) then
                    authHeader.Substring(7)
                else
                    ""

            let jti = getTokenJti token

            let userId =
                if ctx.Items.ContainsKey("UserId") then
                    ctx.Items["UserId"] :?> Guid
                else
                    Guid.Empty

            match jti with
            | Some j ->
                let! exists = tokenRepo.ExistsJti j

                if not exists then
                    let revokedEntity: RevokedTokenEntity =
                        { Id = Guid.NewGuid()
                          Jti = j
                          UserId = userId
                          RevokedAt = DateTime.UtcNow }

                    let! _ = tokenRepo.Create revokedEntity
                    ()
            | None -> ()

            let! activeTokens = rtRepo.ListActiveByUser userId

            for rt in activeTokens do
                let! _ = rtRepo.Update { rt with Revoked = true }
                ()

            return! json {| message = "All sessions logged out" |} next ctx
        }

File: apps/crud-be-java-vertx/src/main/java/com/demobejavx/handler/AuthHandler.java
package com.demobejavx.handler;

import com.auth0.jwt.exceptions.JWTVerificationException;
import com.demobejavx.auth.JwtService;
import com.demobejavx.auth.PasswordService;
import com.demobejavx.contracts.AuthTokens;
import com.demobejavx.contracts.LoginRequest;
import com.demobejavx.contracts.RefreshRequest;
import com.demobejavx.contracts.RegisterRequest;
import com.demobejavx.contracts.User;
import com.demobejavx.domain.model.TokenRevocation;
import com.demobejavx.domain.validation.DomainException;
import com.demobejavx.domain.validation.UserValidator;
import com.demobejavx.domain.validation.ValidationException;
import com.demobejavx.repository.TokenRevocationRepository;
import com.demobejavx.repository.UserRepository;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import io.vertx.core.Future;
import io.vertx.core.Handler;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.web.RoutingContext;
import java.time.Instant;

public class AuthHandler implements Handler<RoutingContext> {

    private static final int MAX_FAILED_ATTEMPTS = 5;
    private static final ObjectMapper MAPPER = new ObjectMapper()
            .registerModule(new JavaTimeModule())
            .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);

    private final UserRepository userRepo;
    private final TokenRevocationRepository revocationRepo;
    private final JwtService jwtService;
    private final PasswordService passwordService;
    private final String action;

    public AuthHandler(String action, UserRepository userRepo,
            TokenRevocationRepository revocationRepo,
            JwtService jwtService, PasswordService passwordService) {
        this.action = action;
        this.userRepo = userRepo;
        this.revocationRepo = revocationRepo;
        this.jwtService = jwtService;
        this.passwordService = passwordService;
    }

    @Override
    public void handle(RoutingContext ctx) {
        switch (action) {
            case "register" -> handleRegister(ctx);
            case "login" -> handleLogin(ctx);
            case "refresh" -> handleRefresh(ctx);
            case "logout" -> handleLogout(ctx);
            case "logout-all" -> handleLogoutAll(ctx);
            default -> ctx.fail(500);
        }
    }

    private void handleRegister(RoutingContext ctx) {
        JsonObject body = ctx.body().asJsonObject();
        if (body == null) {
            ctx.fail(new ValidationException("body", "Request body is required"));
            return;
        }

        RegisterRequest req;
        try {
            req = MAPPER.readValue(body.encode(), RegisterRequest.class);
        } catch (Exception e) {
            ctx.fail(new ValidationException("body", "Invalid request body"));
            return;
        }

        String username = req.getUsername() != null ? req.getUsername() : "";
        String email = req.getEmail() != null ? req.getEmail() : "";
        String password = req.getPassword() != null ? req.getPassword() : "";

        try {
            UserValidator.validateRegistration(username, email, password);
        } catch (ValidationException e) {
            ctx.fail(e);
            return;
        }

        userRepo.existsByUsername(username)
                .compose(exists -> {
                    if (exists) {
                        return Future.failedFuture(new DomainException(409,
                                "Username already exists"));
                    }
                    String hash = passwordService.hash(password);
                    com.demobejavx.domain.model.User newUser = new com.demobejavx.domain.model.User(
                            null, username, email, username,
                            hash, com.demobejavx.domain.model.User.ROLE_USER,
                            com.demobejavx.domain.model.User.STATUS_ACTIVE, 0, Instant.now());
                    return userRepo.save(newUser);
                })
                .onSuccess(user -> {
                    User resp = buildContractUser(user);
                    sendJson(ctx, 201, resp);
                })
                .onFailure(ctx::fail);
    }

    private void handleLogin(RoutingContext ctx) {
        JsonObject body = ctx.body().asJsonObject();
        if (body == null) {
            ctx.fail(401);
            return;
        }

        LoginRequest req;
        try {
            req = MAPPER.readValue(body.encode(), LoginRequest.class);
        } catch (Exception e) {
            ctx.fail(401);
            return;
        }

        String username = req.getUsername() != null ? req.getUsername() : "";
        String password = req.getPassword() != null ? req.getPassword() : "";

        userRepo.findByUsername(username)
                .compose(userOpt -> {
                    if (userOpt.isEmpty()) {
                        return Future.failedFuture(new DomainException(401,
                                "Invalid credentials"));
                    }
                    com.demobejavx.domain.model.User user = userOpt.get();
                    if (com.demobejavx.domain.model.User.STATUS_INACTIVE.equals(user.status())) {
                        return Future.failedFuture(new DomainException(401,
                                "Account deactivated"));
                    }
                    if (com.demobejavx.domain.model.User.STATUS_DISABLED.equals(user.status())) {
                        return Future.failedFuture(new DomainException(401,
                                "Account disabled"));
                    }
                    if (com.demobejavx.domain.model.User.STATUS_LOCKED.equals(user.status())) {
                        return Future.failedFuture(new DomainException(401,
                                "Account locked"));
                    }
                    if (!passwordService.verify(password, user.passwordHash())) {
                        int attempts = user.failedLoginAttempts() + 1;
                        com.demobejavx.domain.model.User updated =
                                user.withFailedLoginAttempts(attempts);
                        if (attempts >= MAX_FAILED_ATTEMPTS) {
                            updated = updated.withStatus(
                                    com.demobejavx.domain.model.User.STATUS_LOCKED);
                        }
                        return userRepo.update(updated)
                                .compose(u -> Future.failedFuture(
                                        new DomainException(401, "Invalid credentials")));
                    }
                    com.demobejavx.domain.model.User resetUser =
                            user.withFailedLoginAttempts(0);
                    return userRepo.update(resetUser);
                })
                .compose(user -> {
                    JwtService.TokenPair tokens = jwtService.generateTokenPair(user);
                    return Future.succeededFuture(tokens);
                })
                .onSuccess(tokens -> {
                    AuthTokens resp = new AuthTokens()
                            .accessToken(tokens.accessToken())
                            .refreshToken(tokens.refreshToken())
                            .tokenType("Bearer");
                    sendJson(ctx, 200, resp);
                })
                .onFailure(ctx::fail);
    }

    private void handleRefresh(RoutingContext ctx) {
        JsonObject body = ctx.body().asJsonObject();
        if (body == null) {
            ctx.fail(401);
            return;
        }

        RefreshRequest req;
        try {
            req = MAPPER.readValue(body.encode(), RefreshRequest.class);
        } catch (Exception e) {
            ctx.fail(401);
            return;
        }

        String refreshToken = req.getRefreshToken() != null ? req.getRefreshToken() : "";

        JwtService.Claims claims;
        try {
            claims = jwtService.validate(refreshToken);
        } catch (JWTVerificationException e) {
            ctx.fail(new DomainException(401, "Token expired or invalid"));
            return;
        }

        if (!"refresh".equals(claims.type())) {
            ctx.fail(new DomainException(401, "Invalid token type"));
            return;
        }

        revocationRepo.isRevoked(claims.jti())
                .compose(revoked -> {
                    if (revoked) {
                        return Future.failedFuture(new DomainException(401,
                                "Token invalid"));
                    }
                    return userRepo.findById(claims.subject());
                })
                .compose(userOpt -> {
                    if (userOpt.isEmpty()) {
                        return Future.failedFuture(new DomainException(401, "User not found"));
                    }
                    com.demobejavx.domain.model.User user = userOpt.get();
                    if (com.demobejavx.domain.model.User.STATUS_DISABLED.equals(user.status())) {
                        return Future.failedFuture(new DomainException(401,
                                "Account disabled"));
                    }
                    if (!com.demobejavx.domain.model.User.STATUS_ACTIVE.equals(user.status())) {
                        return Future.failedFuture(new DomainException(401,
                                "Account deactivated"));
                    }
                    String uid = user.id();
                    if (uid == null) {
                        return Future.failedFuture(new DomainException(500, "User id is null"));
                    }
                    TokenRevocation revoke = new TokenRevocation(claims.jti(), uid,
                            Instant.now());
                    return revocationRepo.save(revoke).map(ignored -> user);
                })
                .compose(user -> {
                    JwtService.TokenPair tokens = jwtService.generateTokenPair(user);
                    String uid = user.id();
                    if (uid == null) {
                        return Future.failedFuture(new DomainException(500, "User id is null"));
                    }
                    TokenRevocation revocation = new TokenRevocation(
                            tokens.refreshJti(), uid, Instant.now());
                    return revocationRepo.save(revocation).map(ignored -> tokens);
                })
                .onSuccess(tokens -> {
                    AuthTokens resp = new AuthTokens()
                            .accessToken(tokens.accessToken())
                            .refreshToken(tokens.refreshToken())
                            .tokenType("Bearer");
                    sendJson(ctx, 200, resp);
                })
                .onFailure(ctx::fail);
    }

    private void handleLogout(RoutingContext ctx) {
        String authHeader = ctx.request().getHeader("Authorization");
        if (authHeader == null || !authHeader.startsWith("Bearer ")) {
            ctx.response().setStatusCode(200).end();
            return;
        }
        String token = authHeader.substring(7);

        JwtService.Claims claims;
        try {
            claims = jwtService.decode(token);
        } catch (Exception e) {
            ctx.response().setStatusCode(200).end();
            return;
        }

        TokenRevocation revocation = new TokenRevocation(claims.jti(), claims.subject(),
                Instant.now());
        revocationRepo.save(revocation)
                .onSuccess(ignored -> ctx.response().setStatusCode(200).end())
                .onFailure(ctx::fail);
    }

    private void handleLogoutAll(RoutingContext ctx) {
        String userId = ctx.get("userId");
        String jti = ctx.get("jti");

        if (userId == null || jti == null) {
            ctx.fail(400);
            return;
        }
        // Revoke the current access token
        TokenRevocation accessRevocation = new TokenRevocation(jti, userId, Instant.now());
        // Revoke a sentinel entry for "all sessions" to block any existing refresh tokens
        TokenRevocation allRevoke = new TokenRevocation(
                "all-" + userId, userId, Instant.now());
        revocationRepo.save(accessRevocation)
                .compose(ignored -> revocationRepo.save(allRevoke))
                .onSuccess(ignored -> ctx.response().setStatusCode(200).end())
                .onFailure(ctx::fail);
    }

    static User buildContractUser(com.demobejavx.domain.model.User user) {
        User.StatusEnum status;
        switch (user.status()) {
            case com.demobejavx.domain.model.User.STATUS_INACTIVE ->
                status = User.StatusEnum.INACTIVE;
            case com.demobejavx.domain.model.User.STATUS_DISABLED ->
                status = User.StatusEnum.DISABLED;
            case com.demobejavx.domain.model.User.STATUS_LOCKED ->
                status = User.StatusEnum.LOCKED;
            default -> status = User.StatusEnum.ACTIVE;
        }
        return new User()
                .id(user.id() != null ? user.id() : "")
                .username(user.username())
                .email(user.email())
                .displayName(user.displayName() != null ? user.displayName() : "")
                .status(status)
                .roles(java.util.List.of(user.role()));
    }

    static void sendJson(RoutingContext ctx, int statusCode, Object obj) {
        try {
            String json = MAPPER.writeValueAsString(obj);
            ctx.response()
                    .setStatusCode(statusCode)
                    .putHeader("Content-Type", "application/json")
                    .end(json);
        } catch (Exception e) {
            ctx.fail(500);
        }
    }
}
